ocaml/xapi/helpers.ml

```diff
@@ -250,6 +250,26 @@ let make_rpc ~__context rpc : Rpc.response =
     else SSL(SSL.make ~use_stunnel_cache:true (), Pool_role.get_master_address(), !Xapi_globs.https_port) in
   XMLRPC_protocol.rpc ~srcstr:"xapi" ~dststr:"xapi" ~transport ~http rpc
 
+let make_timeboxed_rpc ~__context timeout rpc : Rpc.response =
+  let subtask_of = Ref.string_of (Context.get_task_id __context) in
+  Server_helpers.exec_with_new_task "timeboxed_rpc" ~subtask_of:(Context.get_task_id __context) (fun __context ->
+    let open Xmlrpc_client in
+    let http = xmlrpc ~subtask_of ~version:"1.1" "/" in
+    let task_id = Context.get_task_id __context in
+    let cancel () =
+      let resources = Locking_helpers.Thread_state.get_acquired_resources_by_task task_id in
+      List.iter Locking_helpers.kill_resource resources
+    in
+    Xapi_periodic_scheduler.add_to_queue (Ref.string_of task_id) Xapi_periodic_scheduler.OneShot timeout cancel;
+    let transport =
+      if Pool_role.is_master ()
+      then Unix(Xapi_globs.unix_domain_socket)
+      else SSL(SSL.make ~use_stunnel_cache:true ~task_id:(Ref.string_of task_id) (), Pool_role.get_master_address (), !Xapi_globs.https_port)
+    in
+    let result = XMLRPC_protocol.rpc ~srcstr:"xapi" ~dststr:"xapi" ~transport ~http rpc in
```

Comment:
If the `timeout` is quite small, is it possible that the stunnel process has not been created when the `cancel` function been called, so the `Locking_helpers.kill_resource` won't take effect?

Reply:
You're quite right, it'd be a very similar situation to CA-285840, and it's about the only thing that would functionally be better with the original suggestion of using cohttp. However, I believe it's less bad in this case because we're making a connection _to_ the master rather than from as in CA-285840 (so 1/n as likely), and will only go wrong if we've lost our master (or route to the master), in which case either it comes back (and the stunnel connects with/without error), or we get an emergency_reset_master call in which case xapi restarts.